### PR TITLE
[US164 Task178] Fetch Lead Time for Stories

### PR DESCRIPTION
**Pull request by spark290**
[US164 Task178] Fetch Lead Time for Stories

### CODE REVIEW

**Review by AdvaitChirmule — APPROVED** [review]
Looks good to me

File: lead-time/controller/lead_time_controller.py [hunk 1]
```diff
@@ -1,10 +1,134 @@
...
+
+from pydantic import BaseModel
+from datetime import date
+from library.project.getProjectBySlug import get_project_by_slug
+from library.userStory.get_user_story_history import get_closed_user_stories
 
 router = APIRouter()
 
 
+class TimeRequest(BaseModel):
```
Comment (shahih25) [inline lead-time/controller/lead_time_controller.py]: Keep them as the same folder structure. Move this to models

File: lead-time/controller/lead_time_controller.py [hunk 2]
```diff
@@ -1,10 +1,134 @@
...
+    projectslug: str
+    from_date: str = None
+    to_date: str = None
+
+
+class SimpleCache:
```
Comment (shahih25) [inline lead-time/controller/lead_time_controller.py]: Move this to service.

File: lead-time/controller/lead_time_controller.py [hunk 3]
```diff
@@ -1,10 +1,134 @@
...
+        self.cache.clear()
+
+
+# Create an instance of SimpleCache that can be imported in other files
+cache = SimpleCache()
+
+
 @router.post("/metric/LeadTime")
-def get_lead_time_metric():
+def get_lead_time_metric(
```
Comment (shahih25) [inline lead-time/controller/lead_time_controller.py:12]: Only this method should be in controller.

File: lead-time/controller/lead_time_controller.py [hunk 4]
```diff
@@ -1,10 +1,134 @@
...
+    else:
+        return get_lead_time_details(
+            cache.get(lead_time_request.projectslug),
+            token,
+            lead_time_request.from_date,
+            lead_time_request.to_date
+        )
+
+
+def get_lead_time_details(project_details,
```
Comment (shahih25) [inline lead-time/controller/lead_time_controller.py]: All the methods from here till the end, move them to service package which has a file lead_time_service.py

File: lead-time/controller/lead_time_controller.py (hunk 1 above)
Comment (spark290) [reply lead-time/controller/lead_time_controller.py]: Modified

File: lead-time/controller/lead_time_controller.py (hunk 2 above)
Comment (spark290) [reply lead-time/controller/lead_time_controller.py]: Modified

File: lead-time/controller/lead_time_controller.py (hunk 3 above)
Comment (spark290) [reply lead-time/controller/lead_time_controller.py:12]: Modified

File: lead-time/controller/lead_time_controller.py (hunk 4 above)
Comment (spark290) [reply lead-time/controller/lead_time_controller.py]: Modified

File: lead-time/model/lead_time_model.py [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+router = APIRouter()
+
+
+class TimeRequest(BaseModel):
+    projectslug: str
+    from_date: str = None
+    to_date: str = None
+
+
+def get_lead_time_details(project_details,
```
Comment (shahih25) [inline lead-time/model/lead_time_model.py]: Move these methods to lead_time_service.py

File: lead-time/service/lead_time_service.py [hunk 6]
```diff
@@ -0,0 +1,99 @@
+from fastapi import APIRouter
+
+from datetime import date
+from library.userStory.get_user_story_history import get_closed_user_stories
+
+router = APIRouter()
+
+
+class SimpleCache:
```
Comment (shahih25) [inline lead-time/service/lead_time_service.py]: Move this to separate file.

File: lead-time/model/lead_time_model.py (hunk 5 above)
Comment (shahih25) [reply lead-time/model/lead_time_model.py]: Sorry I think this is duplicate. Remove these methods from here. Keep only in service.

Comment (spark290) [reply lead-time/model/lead_time_model.py]: Modified

File: lead-time/service/lead_time_service.py (hunk 6 above)
Comment (spark290) [reply lead-time/service/lead_time_service.py]: Modified

Comment (shahih25) [reply lead-time/service/lead_time_service.py]: Create a file like this (exactly same): https://github.com/SER516/SER516-Team-Cleveland/blob/main/taigaProject/src/taigaApi/util/SimpleCache.py

Then remove from here and update line: https://github.com/SER516/SER516-Team-Cleveland/pull/127/files#diff-7c93933da17a241d8d50decf9fda4d88afefda59e0051d6da64a61c272abeb23R6

Comment (spark290) [reply lead-time/service/lead_time_service.py]: Modified

**Pull request merged**